cookbooks/equilibrium_grain_size/plugins/equilibrium_grain_size.cc

```diff
@@ -0,0 +1,1750 @@
+                                    / (dislocation_creep_exponent[phase_index] * constants::gas_constant * temperature));
+
+      Assert(energy_term > 0.0,
+             ExcMessage("Energy term has to be positive for dislocation creep. Instead it is: " + std::to_string(energy_term)));
+
+      // If we are past the initialization of the adiabatic profile, use it to
+      // limit viscosity variations due to temperature.
+      if (this->get_adiabatic_conditions().is_initialized())
+        {
+          const double adiabatic_energy_term
+            = exp((dislocation_activation_energy[phase_index] + dislocation_activation_volume[phase_index] * pressure)
+                  / (dislocation_creep_exponent[phase_index] * constants::gas_constant * this->get_adiabatic_conditions().temperature(position)));
+
+          Assert(adiabatic_energy_term > 0.0,
+                 ExcMessage("Adiabatic energy term has to be positive for dislocation creep. Instead it is: " + std::to_string(adiabatic_energy_term)));
+
+          const double temperature_dependence = energy_term / adiabatic_energy_term;
+          if (temperature_dependence > max_temperature_dependence_of_eta)
+            energy_term = adiabatic_energy_term * max_temperature_dependence_of_eta;
+          if (temperature_dependence < 1.0 / max_temperature_dependence_of_eta)
+            energy_term = adiabatic_energy_term / max_temperature_dependence_of_eta;
+        }
+
+      const double strain_rate_dependence = (1.0 - dislocation_creep_exponent[phase_index]) / dislocation_creep_exponent[phase_index];
+
+      const double dislocation_viscosity = std::pow(dislocation_creep_prefactor[phase_index],-1.0/dislocation_creep_exponent[phase_index])
+                                           * std::pow(second_strain_rate_invariant,strain_rate_dependence)
+                                           * energy_term;
+
+      Assert(dislocation_viscosity > 0.0,
+             ExcMessage("Dislocation viscosity has to be positive. Instead it is: " + std::to_string(dislocation_viscosity)));
+
+      return dislocation_viscosity;
+    }
+
+
+
+    template <int dim>
+    double
+    EquilibriumGrainSize<dim>::
+    seismic_Vp (const double      temperature,
+                const double      pressure,
+                const std::vector<double> &compositional_fields,
+                const Point<dim> &/*position*/) const
+    {
+      AssertThrow ((reference_compressibility != 0.0) || use_table_properties,
+                   ExcMessage("Currently only compressible models are supported."));
+
+      double vp = 0.0;
+      if (n_material_data == 1)
+        vp = material_lookup[0]->seismic_Vp(temperature,pressure);
+      else
+        {
+          for (unsigned i = 0; i < n_material_data; i++)
+            vp += compositional_fields[i] * material_lookup[i]->seismic_Vp(temperature,pressure);
+        }
+      return vp;
```

Comment:
It might be easier to read this way:
```suggestion
      if (n_material_data == 1)
        return material_lookup[0]->seismic_Vp(temperature,pressure);
      else
        {
          double vp = 0.0;
          for (unsigned i = 0; i < n_material_data; i++)
            vp += compositional_fields[i] * material_lookup[i]->seismic_Vp(temperature,pressure);
          return vp;
        }
```